pkg/executor/sortexec/topn.go

```diff
@@ -166,76 +268,358 @@ func (e *TopNExec) loadChunksUntilTotalLimit(ctx context.Context) error {
+		// The number of rows we loaded may exceeds total limit, remove greatest rows by Pop.
+		heap.Pop(e.chkHeap)
+		e.chkHeap.droppedRowNum++
+	}
+
+	if err := e.executeTopNWhenNoSpillTriggered(ctx); err != nil {
+		e.resultChannel <- rowWithError{err: err}
+		return
+	}
+
+	if e.spillHelper.isSpillNeeded() {
+		if err := e.executeTopNWhenSpillTriggered(ctx); err != nil {
+			e.resultChannel <- rowWithError{err: err}
+			return
+		}
+	}
+
+	e.generateTopNResults()
+}
+
+// Return true when spill is triggered
+func (e *TopNExec) generateTopNResultsWhenNoSpillTriggered() bool {
+	rowPtrNum := len(e.chkHeap.rowPtrs)
+	for ; e.chkHeap.idx < rowPtrNum; e.chkHeap.idx++ {
+		if e.chkHeap.idx%10 == 0 && e.spillHelper.isSpillNeeded() {
+			return true
+		}
+		e.resultChannel <- rowWithError{row: e.chkHeap.rowChunks.GetRow(e.chkHeap.rowPtrs[e.chkHeap.idx])}
+	}
+	return false
+}
+
+func (e *TopNExec) generateResultWithMultiWayMerge(offset int64, limit int64) error {
+	multiWayMerge := newMultiWayMerger(&diskSource{sortedRowsInDisk: e.spillHelper.sortedRowsInDisk}, e.lessRow)
+
+	err := multiWayMerge.init()
+	if err != nil {
+		return err
+	}
+
+	outputRowNum := int64(0)
+	for {
+		if limit != -1 && outputRowNum >= limit {
+			return nil
+		}
+
+		row, err := multiWayMerge.next()
+		if err != nil {
+			return err
+		}
+
+		if row.IsEmpty() {
+			return nil
+		}
+
+		if outputRowNum >= offset {
+			select {
+			case <-e.finishCh:
+				return nil
+			case e.resultChannel <- rowWithError{row: row}:
+			}
+		}
+		outputRowNum++
+		injectParallelSortRandomFail(1)
 	}
-	newRowChunks.GetMemTracker().SetLabel(memory.LabelForRowChunks)
-	e.memTracker.ReplaceChild(chkHeap.rowChunks.GetMemTracker(), newRowChunks.GetMemTracker())
-	chkHeap.rowChunks = newRowChunks
+}
+
+// Generate results with this function when we trigger spill only once.
+// It's a public function as we need to test it in ut.
+func (e *TopNExec) GenerateTopNResultsWhenSpillOnlyOnce() error {
+	inDisk := e.spillHelper.sortedRowsInDisk[0]
+	chunkNum := inDisk.NumChunks()
+	skippedRowNum := uint64(0)
+	offset := e.Limit.Offset
+	for i := 0; i < chunkNum; i++ {
+		chk, err := inDisk.GetChunk(i)
+		if err != nil {
+			return err
+		}
+
+		injectTopNRandomFail(10)
 
-	e.memTracker.Consume(int64(8 * (len(newRowPtrs) - len(chkHeap.rowPtrs))))
-	chkHeap.rowPtrs = newRowPtrs
+		rowNum := chk.NumRows()
+		j := 0
+		if !e.inMemoryThenSpillFlag {
+			// When e.inMemoryThenSpillFlag == false, we need to manually set j
+			// because rows that should be ignored before offset have also been
+			// spilled to disk.
+			if skippedRowNum < offset {
+				rowNumNeedSkip := offset - skippedRowNum
+				if rowNum > int(rowNumNeedSkip) {
+					j += int(rowNumNeedSkip)
+					skippedRowNum += rowNumNeedSkip
+				} else {
+					// All rows in this chunk should be skipped
+					skippedRowNum += uint64(rowNum)
+					continue
+				}
+			}
+		}
+
+		for ; j < rowNum; j++ {
```

Comment:
limit is not checked here because the spill stage guarantees that at most `offset + limit` rows will be spilled for each spill?

Reply:
> limit is not checked here because the spill stage guarantees that at most `offset + limit` rows will be spilled for each spill?

yes